pyproject.toml

```diff
@@ -0,0 +1,92 @@
+[build-system]
+requires = [
+    "hatchling>=1.13.0",
+    "hatch-vcs>=0.3.0",
+]
+build-backend = "hatchling.build"
+
+[project]
+name = "pyshaper"
+dynamic = ["version"]
```

Comment:
A normal thing to do is to determine the version of a library dynamically from version control when the wheel is being built. This also allow for provenance to be established more clearly.

Example:

For the current state of this PR

```console
$ python -m pip show pyshaper
Name: pyshaper
Version: 0.1.dev49
Summary: framework for defining, building, and evaluating generalized shape observables for collider physics
Home-page: 
Author: 
Author-email: Rikab Gambhir <rikab@mit.edu>
License: MIT
Location: /home/feickert/.pyenv/versions/3.10.6/envs/pyshaper-dev/lib/python3.10/site-packages
Editable project location: /home/feickert/Code/GitHub/forks/SHAPER
Requires: geomloss, matplotlib, numpy, pyjet, scipy, torch
Required-by: 
```

the version is `0.1.dev49` because there are no tags created for the repository so it doesn't know what version it should be yet (it thinks it is `0.0.0`) and it assumes that the next version will be `0.1.0` and as there are 49 commits in the history of the repo up to this point in the PR

```console
$ git rev-list --count HEAD
49
```

then it thinks we're at the 49th developer commit towards `v0.1.0`.

A bit of a problem is that the repository has been branded as `v1.0.0`

https://github.com/rikab/SHAPER/blob/5a7a98870c36be9a96bc22d5d66075b3a35774da/README.md?plain=1#L1

https://github.com/rikab/SHAPER/blob/5a7a98870c36be9a96bc22d5d66075b3a35774da/README.md?plain=1#L50

without ever making anything resembling a release in terms of a [tag](https://github.com/rikab/SHAPER/tags) or a [GitHub release](https://github.com/rikab/SHAPER/releases). Also a major release in SemVer (`v1.0.0`) indicates ([culturally and historically, but not in a binding sense](https://hynek.me/articles/semver-will-not-save-you/)) that the API is going to be stable moving forward (to some degree) until a `v2.0.0` API is made. If you're fine with this and don't plan on doing any further development then that's great, but if you think you might be wanting to make some breaking changes to the API in the near future I would recommend instead starting with `v0.1.0` and working your way up towards a `v1.0.0` stable release through many smaller SemVer releases.

Reply:
I'm not sure I completely follow here -- It's not clear to me what tags or Github release do, or how this is separate from making the repo public. I don't plan on making further development to the API (at least, in the near future), other than patches as the need arises (and, as you suggested elsewhere, updating the pyjet dependency to fastjet).

Reply:
> It's not clear to me what tags or Github release do, or how this is separate from making the repo public

Making the repo public doesn't do anything other then let people see your source code.

At the moment there's nothing that makes this repo a `v1.0.0` other than you saying that it is in text 

https://github.com/rikab/SHAPER/blob/5a7a98870c36be9a96bc22d5d66075b3a35774da/README.md?plain=1#L50

As is, without tags, there's no (reasonable) way to programmatically determine if there is a new version of the source code that would constitute a new release. A very conventional way to do this is to use tags in the version control system, which allows everyone to agree on what is the exact commit hash that people should be using when they want `vX.Y.Z` without needing to explicitly give the full hash &mdash; they use the tag as an alias. In Git this is done just with [`git tag`](https://git-scm.com/book/en/v2/Git-Basics-Tagging)

```console
# example
$ git tag v1.0.0
```

Tags are very versatile and there could be situations in which you might want to create multiple tags at the same commit or you might want to tag a commit but not create a new release from it. GitHub offers the concept of a "GitHub release" which creates a human readable page from a particular tag you select relative to the last GitHub release. Multiple services use GitHub releases as webhooks to trigger automatic systems, like [creating a new Zenodo archive](https://docs.github.com/en/repositories/archiving-a-github-repository/referencing-and-citing-content#issuing-a-persistent-identifier-for-your-repository-with-zenodo) or [creating a new distribution of a release in CI and deploying it to distribution sites](https://github.com/scikit-hep/pyhf/blob/4377bdf461816c7c5c5d0e42a2d61c3afb9a0858/.github/workflows/publish-package.yml#L11-L12) (like PyPI).

For more on GitHub's relationship with tags and releases c.f. [their docs on releasing projects](https://docs.github.com/en/repositories/releasing-projects-on-github/about-releases).

While it is possible to statically define version information that would correspond to the Git tag in your metadata and packaging configuration file (here `pyproject.toml`), and sometimes has advantages, this can sometimes lead to discrepancies in provenance RE: what is the commit for a particular release. Most packaging build backends (like `hatchling` (used here), or `setuptools`) have systems built into them or around them that allow for the version of the code to be determined from the VCS itself.

Using this PR as an explicit example, as there are no tags associated to the Git repo yet [`importlib`](https://docs.python.org/3/library/importlib.metadata.html) shows that the library version a developer release of `v0.1.0` (what it expects to be the next tag)

```console
$ python -m pip install --upgrade -e .  # Creating an editable install of the local sorce
$ python -c 'from importlib.metadata import version; print(version("pyshaper"))'                           
0.1.dev50
```

As a local example, if on my machine if I now say that I think this should be tag `v0.2.0`, then `importlib` understands this from the metadata of the install

```console
$ git tag v0.2.0
$ python -m pip install --upgrade -e .
$ python -c 'from importlib.metadata import version; print(version("pyshaper"))'                           
0.2.0
$ python -m pip show pyshaper
Name: pyshaper
Version: 0.2.0
Summary: framework for defining, building, and evaluating generalized shape observables for collider physics
Home-page: 
Author: 
Author-email: Rikab Gambhir <rikab@mit.edu>
License: MIT
Location: /home/feickert/.pyenv/versions/3.10.6/envs/pyshaper-dev/lib/python3.10/site-packages
Editable project location: /home/feickert/Code/GitHub/forks/SHAPER
Requires: geomloss, matplotlib, numpy, pyjet, scipy, torch
Required-by:
```

Let me know if this makes sense.

---

I would suggest doing the (VERY MUCH OPTIONAL) following actions before creating a tag for the public release to add additional functionality around it (these operations can be done independently of this PR):
* Create a Zenodo account and [setup the GitHub integration](https://docs.github.com/en/repositories/archiving-a-github-repository/referencing-and-citing-content#issuing-a-persistent-identifier-for-your-repository-with-zenodo) which allows for automatically creating a Zenodo archive of your code and an associated DOI for it whenever you make a GitHub release.
* Create a [CITATION.cff file](https://citation-file-format.github.io/) for the repository that allows for programmatic generation of citation information _for the code_ in the repository.
   - Additionally, expand upon the arXiv citation information to make the [code citation information clear in the README as well](https://github.com/scikit-hep/pyhf/blob/4377bdf461816c7c5c5d0e42a2d61c3afb9a0858/README.rst#citation) (even though GitHub will already make a little citation button for you).
   - c.f. https://github.com/scikit-hep/vector/pull/243 for an example as well as lots of linked PRs.

Doing these would mean that creating of a tag and GitHub release that correspond to the public release you wanted will additionally have the code clearly citeable in a way you intend and archived on Zenodo.

This is a lot of text, and probably too many new ideas for a single GitHub comment. I'm happy to break these apart and discuss them with you in separate discussions in Issues.

Reply:
Ah ok, I think I now understand how/why the versioning system works. I'll do these steps you recommended.

Reply:
Sounds good. If you have any questions on them though please feel free to ask me (probably easiest to open up a GitHub Issue and just tag me)!

Reply:
I think I managed to get the tag and official release done. Following the packaging tutorial you sent me, it seems the only thing left for me to do is to upload to PyPi, which I'll try my hand at.

Reply:
@rikab actually here's that comment translated for this project:

---

Okay, so here's my suggestions of steps for deploying to TestPyPI and PyPI manually

## Get accounts

1. [Register for account on TestPyPI](https://test.pypi.org/account/register/)
2. [Register for account on PyPI](https://pypi.org/account/register/)


## Setup virtual environment for deployment

1. Activate your virtual environment (or create a new one if needed)
2. Update `pip`, `setuptools`, `wheel`

```
python -m pip install --upgrade pip setuptools wheel
```

3. Install `build` and `twine`

```
python -m pip install --upgrade build twine
```

## Deploy tag `v1.0.0`

There already exists the tag on GitHub `v1.0.0` which you just made. However, that isn't the current `HEAD` of the `main` branch, so to get the correct version in the distributions we can rewind and tag the same commit with a `v1.0.0` tag and deploy _that_.

1. For simplicity, clone the project fresh from GitHub

```
cd /tmp
git clone git@github.com:rikab/SHAPER.git
cd SHAPER
```

2. Reset to the state of the `v1.0.0` tag

```
git reset --hard v1.0.0
```

4. Build a sdist and wheel distribution with `build`

```console
$ python -m build .
...
Successfully built pyshaper-1.0.0.tar.gz and pyshaper-1.0.0-py3-none-any.whl
```

5. Check the state of the distributions with `twine`

```console
$ twine check dist/*
Checking dist/pyshaper-1.0.0-py3-none-any.whl: PASSED
Checking dist/pyshaper-1.0.0.tar.gz: PASSED
```

6. List the contents of the sdist (which is a tarball) and the wheel (which is actually a zipfile) to manually inspect them

```
python -m tarfile --list dist/pyshaper-*.tar.gz
```
```
python -m zipfile --list dist/pyshaper-*.whl
```

The wheel will be a subset of the sdist (_source_ distribution) that just contains the actual package files.

7. Upload the distributions to TestPyPI (this will require your TestPyPI username and password from before)

```
twine upload --repository testpypi dist/*
```

9. Check that things look good on TestPyPI: https://test.pypi.org/project/pyshaper/ :+1:

Assuming they do...

10. Upload the distributions to PyPI (this will require your PyPI username and password from before)

```
twine upload dist/*
```

Congratulations! You have just released a project to the Python Package Index! https://pypi.org/project/pyshaper/ :tada:

## Register API tokens for your project

Now that you have gone through the very manual first time procedure you can setup API tokens to make this easier in the future.

### TestPyPI

1. Login to [TestPyPI](https://test.pypi.org/manage/account/token/)
2. Create an [API token](https://pypi.org/help/#apitoken) for your project.
   * [Account settings -> API tokens](https://test.pypi.org/manage/account/token/)
      - **Token name:** pyshaper deployment (or something similarly descriptive)
      - **Scope:** Project: pyshaper
   * **Save the token value to your password manager and note this is for TestPyPI**

### PyPI

1. Login to [PyPI](https://test.pypi.org/manage/account/token/)
2. Create an [API token](https://pypi.org/help/#apitoken) for your project.
   * [Account settings -> API tokens](https://test.pypi.org/manage/account/token/)
      - **Token name:** pyshaper deployment (or something similarly descriptive)
      - **Scope:** Project: pyshaper
   * **Save the token value to your password manager and note this is for PyPI**

## General deployment strategy in the future

I have a totally CI based workflow that I can share in the future that allows for deploying to TestPyPI and PyPI with [just a few button clicks](https://www.youtube.com/watch?v=tsAWDumpcW8). Here's a manual workflow though.

1. From the commit that you want to release from (probably the latest commit on GitHub (`origin/master`)) bump the version information in your repository to the SemVer version you want and also create a corresponding tag at that commit.

You can use a tool like [`tbump`](https://github.com/your-tools/tbump) to do this.

2. Push that commit with the new versions and tag to GitHub

```
git push origin HEAD --tags
```

3. Build a sdist and wheel distribution with `build`

```
rm -rf build
python -m build .
```

4. Check the state of the distributions with `twine`

```
twine check dist/*
```

5. List the contents of the sdist (which is a tarball) and the wheel (which is actually a zipfile) to manually inspect them

```
python -m tarfile --list dist/pyshaper-*.tar.gz
```
```
python -m zipfile --list dist/pyshaper-*.whl
```

6. Upload the distributions to TestPyPI (this will require your TestPyPI username and API token)

```
twine upload --repository testpypi dist/*
```

7. Check that things look good on TestPyPI: https://test.pypi.org/project/pyshaper/ :+1:

Assuming they do...

8. Upload the distributions to PyPI (this will require your PyPI username and API token)

```
twine upload dist/*
```

9. Create a [GitHub release](https://github.com/rikab/SHAPER/releases) from the same tag you just deployed to PyPI to have corresponding releases across the platforms.